datashader/glyphs/glyph.py

```diff
@@ -93,11 +95,20 @@ def _compute_bounds_2d(vals):
 
     @staticmethod
     def to_cupy_array(df, columns):
-        if cudf.__version__ >= Version("22.02"):
+        if isinstance(columns, tuple):
+            columns = list(columns)
+
+        # cuDF does not extract the same column name multiple
+        # times but only one time. Where pandas extract the
+        # column name multiple times. See PR 1050 for more details.
+        if isinstance(columns, list) and (len(columns) != len(set(columns))):
+            return cp.stack([cp.array(df[c]) for c in columns], axis=1)
```

Comment:
`df` will still contain only unique column series (the earlier added one is just deleted), so you are essentially creating a `cp.stack` with two of the rows as same values. in this case, say duplicate column `x`, calling `df['x']` two times, because the `columns` list contains it twice, will still end up calling the same `df['x']` twice.

Reply:
Exactly, we want to return the same column twice in a `cupy` array. This is why the tests in my first post were failing for `cudf` but not for `pandas`.

Reply:
In that case, wouldn't it be better to fix the test to pass `unique(y_names)` for cudf-tests, instead of adding it here? 

What exact test is failing, can you help me find the exact test which has duplicate column names?

Reply:
The tests failing are not testing this function directly, but it uses this function at some point to extract an array from a DataFrame. The arrays are then used for further calculation, which is sensitive to if the output is a 1D array (with unique / without stack) or a 2D array (without unique / with stack). 

Failing tests for 22.02:
``` python
FAILED datashader/datashader/tests/test_pandas.py::test_line_manual_range[df_args5-cvs_kwargs5-cudf_DataFrame]
FAILED datashader/datashader/tests/test_pandas.py::test_line_autorange[df_args4-cvs_kwargs4-cudf_DataFrame]
FAILED datashader/datashader/tests/test_pandas.py::test_area_to_zero_autorange[df_kwargs4-cvs_kwargs4-cudf_DataFrame]
FAILED datashader/datashader/tests/test_pandas.py::test_area_to_line_autorange[df_kwargs4-cvs_kwargs4-cudf_DataFrame]
```


<details>
  <summary>A failed tests output (copied from the log files in the first post) </summary>


``` python
______ test_area_to_zero_autorange[df_kwargs4-cvs_kwargs4-cudf_DataFrame] ______

DataFrame = <function cudf_DataFrame at 0x7f997d1b90d0>
df_kwargs = {'data': {'x0': [-4, -2, 0], 'x1': [0, 2, 4], 'y0': [0, -4, 0]}, 'dtype': 'float32'}
cvs_kwargs = {'axis': 0, 'x': ['x0', 'x1'], 'y': 'y0'}

    @pytest.mark.parametrize('DataFrame', DataFrames)
    @pytest.mark.parametrize('df_kwargs,cvs_kwargs', [
        # axis1 none constant
        (dict(data={
            'x0': [-4, 0],
            'x1': [-2, 2],
            'x2': [0, 4],
            'y0': [0, 0],
            'y1': [-4, -4],
            'y2': [0, 0]
        }, dtype='float32'), dict(x=['x0', 'x1', 'x2'], y=['y0', 'y1', 'y2'], axis=1)),
    
        # axis1 y constant
        (dict(data={
            'x0': [-4, 0],
            'x1': [-2, 2],
            'x2': [0, 4],
        }, dtype='float32'),
         dict(x=['x0', 'x1', 'x2'], y=np.array([0, -4, 0], dtype='float32'), axis=1)),
    
        # axis0 single
        (dict(data={
            'x': [-4, -2, 0, 0, 2, 4],
            'y': [0, -4, 0, 0, -4, 0],
        }), dict(x='x', y='y', axis=0)),
    
        # axis0 multi
        (dict(data={
            'x0': [-4, -2, 0],
            'x1': [0, 2, 4],
            'y0': [0, -4, 0],
            'y1': [0, -4, 0],
        }, dtype='float32'), dict(x=['x0', 'x1'], y=['y0', 'y1'], axis=0)),
    
        # axis0 multi, y string
        (dict(data={
            'x0': [-4, -2, 0],
            'x1': [0, 2, 4],
            'y0': [0, -4, 0],
        }, dtype='float32'), dict(x=['x0', 'x1'], y='y0', axis=0)),
    
        # axis1 ragged arrays
        (dict(data={
            'x': pd.array([[-4, -2, 0], [0, 2, 4]], dtype='Ragged[float32]'),
            'y': pd.array([[0, -4, 0], [0, -4, 0]], dtype='Ragged[float32]')
        }), dict(x='x', y='y', axis=1))
    ])
    def test_area_to_zero_autorange(DataFrame, df_kwargs, cvs_kwargs):
        if cudf and DataFrame is cudf_DataFrame:
            if isinstance(getattr(df_kwargs['data'].get('x', []), 'dtype', ''), RaggedDtype):
                pytest.skip("cudf DataFrames do not support extension types")
    
        df = DataFrame(**df_kwargs)
    
        axis = ds.core.LinearAxis()
        lincoords_y = axis.compute_index(
            axis.compute_scale_and_translate((-4., 0.), 7), 7)
        lincoords_x = axis.compute_index(
            axis.compute_scale_and_translate((-4., 4.), 13), 13)
    
        cvs = ds.Canvas(plot_width=13, plot_height=7)
    
        agg = cvs.area(df, agg=ds.count(), **cvs_kwargs)
    
        sol = np.array([[0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
                        [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
                        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
                        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
                        [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0],
                        [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0],
                        [1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1]],
                       dtype='i4')
    
        out = xr.DataArray(sol, coords=[lincoords_y, lincoords_x],
                           dims=['y', 'x'])
>       assert_eq_xr(agg, out)

datashader/datashader/tests/test_pandas.py:1548: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

agg = <xarray.DataArray (y: 7, x: 13)>
array([[0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 1, 0, 0, 0,...077 -2.462 -1.846 ... 1.846 2.462 3.077 3.692
  * y        (y) float64 -3.714 -3.143 -2.571 -2.0 -1.429 -0.8571 -0.2857
b = <xarray.DataArray (y: 7, x: 13)>
array([[0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 1,...143 -2.571 -2.0 -1.429 -0.8571 -0.2857
  * x        (x) float64 -3.692 -3.077 -2.462 -1.846 ... 1.846 2.462 3.077 3.692
close = False

    def assert_eq_xr(agg, b, close=False):
        """Assert that two xarray DataArrays are equal, handling the possibility
        that the two DataArrays may be backed by ndarrays of different types"""
        if cupy:
            if isinstance(agg.data, cupy.ndarray):
                agg = xr.DataArray(
                    cupy.asnumpy(agg.data), coords=agg.coords, dims=agg.dims
                )
            if isinstance(b.data, cupy.ndarray):
                b = xr.DataArray(
                    cupy.asnumpy(b.data), coords=b.coords, dims=b.dims
                )
        if close:
            xr.testing.assert_allclose(agg, b)
        else:
>           xr.testing.assert_equal(agg, b)
E           AssertionError: Left and right DataArray objects are not equal
E           
E           Differing values:
E           L
E               array([[0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
E                      [0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
E                      [0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0],
E                      [0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0],
E                      [0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0],
E                      [0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0],
E                      [1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 0, 0, 0]], dtype=uint32)
E           R
E               array([[0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
E                      [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
E                      [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
E                      [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
E                      [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0],
E                      [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0],
E                      [1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1]], dtype=int32)

datashader/datashader/tests/test_pandas.py:104: AssertionError

```

</details>

The output fails because the two final arrays do not match with the expected solution e.g. the last value of the array should be one but is zero.

Reply:
tried to run it locally, now I get what you are trying to say. scenarios where `cvs.line` has argumnets like `x=['x0', 'x1'], y='y0'`, basically single `y` axis, it gets to the function as `[y0, y0]`, and cuDF, as you rightly pointed out, instead of returning a 2d array returns a 1d array.

Okay, in that case, this solution works! 

Thanks a lot for digging into this issue and fixing it, really appreciate it!!

Reply:
tried to run it locally, now I get what you are trying to say. scenarios where `cvs.line` has arguments like `x=['x0', 'x1'], y='y0'`, basically single `y` axis, it gets to the function as `[y0, y0]`, and cuDF, as you rightly pointed out, instead of returning a 2d array returns a 1d array.

Okay, in that case, this solution works! 

Thanks a lot for digging into this issue and fixing it, really appreciate it!!